File: test_src/fortran/hierarchical_parallelism/atomic/target_parallel.f90

program target_parallel
    USE OMP_LIB

    implicit none
  
    
    REAL :: COUNTER = 0

    
    
    INTEGER :: num_threads
    

     
    
    !$OMP TARGET PARALLEL   MAP(TOFROM: COUNTER) 



    
    num_threads = omp_get_num_threads()
    

    

!$OMP ATOMIC UPDATE

counter =  counter +1./num_threads


 
     

    !$OMP END TARGET PARALLEL
    

    IF  ( ( ABS(COUNTER - 1) ) > 10*EPSILON(COUNTER) ) THEN
        write(*,*)  'Expected 1 Got', COUNTER
        call exit(1)
    ENDIF

end program target_parallel

File: test_src/cpp/hierarchical_parallelism/atomic/float/target_parallel.cpp
#include <iostream>
#include <limits>
#include <cmath>




#ifdef _OPENMP
#include <omp.h>
#else
int omp_get_num_teams() {return 1;}
int omp_get_num_threads() {return 1;}
#endif


bool almost_equal(float x, float y, int ulp) {

     return std::fabs(x-y) <= std::numeric_limits<float>::epsilon() * std::fabs(x+y) * ulp ||  std::fabs(x-y) < std::numeric_limits<float>::min();

}

void test_target_parallel.cpp(){

 // Input and Outputs
 

float counter{};

// Main program

#pragma omp target parallel  map(tofrom:counter) 

{

const int num_threads = omp_get_num_threads();



#pragma omp atomic update

counter += float { 1.0f } / num_threads ;



}


// Validation
if ( !almost_equal(counter,float { 1 }, 10)  ) {
    std::cerr << "Expected: " << 1 << " Got: " << counter << std::endl;
    throw std::runtime_error( "target_parallel.cpp give incorect value when offloaded");
}

}
int main()
{
    test_target_parallel.cpp();
}
